File: src/templates/scripts/action_server_queue_goals.py
#!/usr/bin/env python3
import rospy
import actionlib
import threading
from custom_msgs.msg import ExampleAction, ExampleResult, ExampleFeedback

class TemplateActionServer():

    def __init__(self):
        self.action_server = actionlib.ActionServer(
            ns='/example_action',
            ActionSpec=ExampleAction,
            goal_cb=self.goal_callback,
            cancel_cb=self.cancel_callback,
            auto_start=False
        )
        self.action_server.start()

        self.cancel_goals = {}

        self.goal_queue = []
        w = threading.Thread(name="worker", target=self.run_queue)
        w.start()

        rospy.loginfo(msg='Action server is in standby')

    def execute_callback(self, goal_handle):
        rospy.loginfo(msg='Executing new goal')

        goal = goal_handle.get_goal().goal
        result = ExampleResult()
        feedback = ExampleFeedback()
        success = False
        cancel = False

        goal_id = goal_handle.get_goal_id().id

        rate = rospy.Rate(1.0)

        counter = 0
        max_num = 10

        while True:

            if self.cancel_goals[goal_id]:
                cancel = True
                break

            if counter == goal:
                success = True
                break

            if counter >= max_num:
                break

            counter += 1

            feedback.feedback = counter
            goal_handle.publish_feedback(feedback=feedback)

            rate.sleep()

        result.result = success

        if cancel:
            goal_handle.set_canceled(result)
            rospy.logwarn(msg='Goal canceled')
        elif success:
            goal_handle.set_succeeded(result)
            rospy.loginfo(msg='Goal succeeded')
        else:
            goal_handle.set_aborted(result)
            rospy.logerr(msg='Goal aborted')

        self.cancel_goals.pop(goal_id)

    def run_queue(self):
        rate = rospy.Rate(1.0)

        while not rospy.is_shutdown():
            
            if len(self.goal_queue) > 0:
                self.execute_callback(self.goal_queue.pop(0))
            rate.sleep()

    def goal_callback(self, goal_handle):
        if goal_handle.get_goal().goal > 0:
            goal_handle.set_accepted()
            rospy.loginfo(msg='Incoming goal accepted')
        else:
            goal_handle.set_rejected()
            rospy.logwarn(msg='Incoming goal rejected')
            return
        
        self.goal_queue.append(goal_handle)
        self.cancel_goals[goal_handle.get_goal_id().id] = False

    def cancel_callback(self, goal_handle):
        goal_id = goal_handle.get_goal_id().id

        if goal_id in self.cancel_goals:
            self.cancel_goals[goal_id] = True
            rospy.logwarn(msg='Canceling queue goals')

def main():
    rospy.init_node(name="action_server")
    TemplateActionServer()
    rospy.spin()
    
if __name__ == '__main__':
    main()

File: src/templates/src/action_server_queue_goals.cpp
#include <ros/ros.h>
#include <actionlib/server/action_server.h>
#include <custom_msgs/ExampleAction.h>
#include <thread>
#include <queue>

typedef actionlib::ActionServer<custom_msgs::ExampleAction>::GoalHandle GoalHandle;

class TemplateActionServer
{
    public:

    TemplateActionServer() : action_server
    (
        nh, "/example_action",
        boost::bind(&TemplateActionServer::goal_callback, this, _1),
        boost::bind(&TemplateActionServer::cancel_callback, this, _1),
        false
    )
    {
        action_server.start();

        std::thread(&TemplateActionServer::run_queue, this).detach();

        ROS_INFO("Action server is up");
    }

    void execute_callback(GoalHandle goal_handle)
    {
        ROS_INFO("Executing new goal");

        int goal = goal_handle.getGoal()->goal;
        custom_msgs::ExampleResult result;
        custom_msgs::ExampleFeedback feedback;
        bool success = false;
        bool cancel = false;

        std::string goal_id = goal_handle.getGoalID().id;

        ros::Rate rate(1.0);

        int counter = 0;
        int max_num = 10;

        while (ros::ok())
        {
            if (cancel_goals[goal_id])
            {
                cancel = true;
                break;
            }

            if (counter == goal)
            {
                success = true;
                break;
            }

            if (counter >= max_num)
            {
                break;
            }

            counter++;

            feedback.feedback = counter;
            goal_handle.publishFeedback(feedback);

            rate.sleep();
        }

        result.result = counter;

        if (cancel)
        {
            goal_handle.setCanceled(result);
            ROS_WARN("Goal canceled");
        }
        else if (success)
        {
            goal_handle.setSucceeded(result);
            ROS_INFO("Goal succeeded");
        }
        else
        {
            goal_handle.setAborted(result);
            ROS_ERROR("Goal aborted");
        }

        cancel_goals.erase(goal_id);
    }

    void run_queue()
    {
        ros::Rate rate(1.0);

        while (ros::ok())
        {
            if (goal_queue.size() > 0)
            {
                TemplateActionServer::execute_callback(goal_queue.front());
                goal_queue.pop();
            }

            rate.sleep();
        }
    }

    void goal_callback(GoalHandle goal_handle)
    {
        if (goal_handle.getGoal()->goal > 0)
        {
            goal_handle.setAccepted();
            ROS_INFO("Incoming goal accepted");
        }
        else
        {
            goal_handle.setRejected();
            ROS_INFO("Incoming goal rejected");
            return;
        }

        goal_queue.push(goal_handle);
        cancel_goals[goal_handle.getGoalID().id] = false;
    }

    void cancel_callback(GoalHandle goal_handle)
    {
        std::string goal_id = goal_handle.getGoalID().id;

        if (cancel_goals.find(goal_id) != cancel_goals.end())
        {
            cancel_goals[goal_id] = true;
            ROS_WARN("Canceling goal");
        }
    }

    private:

    ros::NodeHandle nh;
    actionlib::ActionServer<custom_msgs::ExampleAction> action_server;
    std::map<std::string, bool> cancel_goals {};
    std::queue<GoalHandle> goal_queue;
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "action_server");
    TemplateActionServer action_server;
    ros::spin();

    return 0;

}